Sprint 2/step 1

yuriydyuster opened the pull request:
См. комментарий к последнему коммиту

yuriydyuster committed:
Обновлено:
 - список ингридитентов перенесен в контекст (но еще не везде используется от туда)
 - добавлена реакция на нажатие правой кнопкой мыши по карточке ингридиента, так пока они будут добавляться в заказ для тиестирования

yuriydyuster committed:
Обновлено:
 - обновлен файл README.md

yuriydyuster committed:
Обновлено:
 - нажатие правой кнопкой мыши на карточке ингрединета добавляет его в конструктор бургера;
 - список доступных ингрендиентов и список выбранных ингредиентов перенесены в контекст.

yuriydyuster committed:
Реализовано:
 - (временно, для тестирования) нажатие правой кнопкой мыши на карточке ингрединета добавляет его в конструктор бургера;
 - список доступных ингрендиентов, список заказанных ингредиентов, общая сумма заказа перенесены в контекст;
 - добавленны сервисные функции getIngredientById, isBunOrdered;
 - добавление/удаление ингредиента в заказ реализованно с помощью хука useReducer с учетом обозначенных ограничений, процесс пока логируется в консоль;
 - отправка заказа реализована через предоставленное API;
 - времмено ведуться логи в консоль, чтобы лучше понимать, что происходит.

besovadevka commented on src/components/services/app-context.tsx:
```diff
@@ -0,0 +1,41 @@
+
+export const API_URL = 'https://norma.nomoreparties.space/api/';
+
+export const IngredientsContext = React.createContext({
+    ingredients: [],
+    isLoading: false
+});
+
+
+export const OrderContext = React.createContext({
+    orderList: [],
+    orderListDispatcher: undefined
+});
+
+export const PriceContext = React.createContext({
+    totalPrice: 0,
+    setTotalPrice: undefined
+});
```
Супер, что контексты объявлены в отдельной папке `services`

besovadevka commented on src/components/services/app-context.tsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
```
Только папка `services` должна находиться в `src`, а не в папке с компонентами)

besovadevka commented on src/components/services/app-context.tsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import {IngredientCardProps} from "../ingredient-card/ingredient-card";
+
+export const API_URL = 'https://norma.nomoreparties.space/api/';
```
Константу также можно вынести в отдельный файл, например, `constants.ts`

besovadevka commented on src/components/services/app-context.tsx:
```diff
@@ -0,0 +1,41 @@
+
+export function isBunOrdered (ingredients: IngredientCardProps[], orderList: string[]) : boolean {
+
+    let result = false;
+    orderList.forEach((item) => {
+        const current = getIngredientByID(ingredients, item);
+        if (current) result = result || (current.type === "bun");
+    })
```
Тут также можно воспользоваться методом `some`, он позволяет получить сразу булевое значение, если один из элементов массива в коллбэке вернул положительный ответ
```suggestion
    const isBunOrdered = orderList.some((item) => getIngredientByID(ingredients, item)?.type === "bun";
    })
```

besovadevka commented on src/components/burger-ingredients/burger-ingredients.tsx:
```diff
@@ -41,8 +41,11 @@ const IngredientsTabs = (props: {
         document.location.href = "#"+current;
     }, [current]);
 
+    const {ingredients, isLoading} = useContext(IngredientsContext);
+    const {orderList} = useContext(OrderContext);
```
Отличное применение контекста)

besovadevka commented on src/components/app/App.tsx:
```diff
@@ -1,63 +1,116 @@
-        '60d3b41abdacab0026a733c6'
-    ]);
+    const initialOrderList : string[] = [
+        //'60d3b41abdacab0026a733c6',
+        // '60d3b41abdacab0026a733d0',
+        // '60d3b41abdacab0026a733cb',
+        // '60d3b41abdacab0026a733d3',
+        // '60d3b41abdacab0026a733d4',
+        // '60d3b41abdacab0026a733cc',
+        //'60d3b41abdacab0026a733c6'
```
Тут тоже, так как реализована функциональность добавления ингредиентов, массив можно оставить пустым

besovadevka commented on src/components/app/App.tsx:
```diff
@@ -1,63 +1,116 @@
     return (
-        <>
+        // @ts-ignore
+        <IngredientsContext.Provider value={{ingredients, isLoading}}>
```
В контекст нужно обернуть только те, компоненты, которые его используют

besovadevka commented on src/components/app/App.tsx:
```diff
@@ -1,63 +1,116 @@
-                    <BurgerIngredients ingredientList={ingredients.data} orderList={orderList}/>
-                    <BurgerConstructor orderList={orderList} reConstructBurger={setOrderList} ingredients={ingredients.data}/>
+                    {/*// @ts-ignore*/}
+                    <OrderContext.Provider value={{orderList, orderListDispatcher, totalPrice, setTotalPrice}} >
```
Кажется, тут пробрасываются лишние аргументы

besovadevka commented on src/components/app/App.tsx:
```diff
@@ -1,63 +1,116 @@
+                    {/*// @ts-ignore*/}
+                    <OrderContext.Provider value={{orderList, orderListDispatcher, totalPrice, setTotalPrice}} >
+                        <BurgerIngredients />
+                        {/*// @ts-ignore*/}
```
Чтобы убрать эти ошибки, нужно при создании контекста указать тип его содержимого)

besovadevka commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -1,81 +1,135 @@
         if (current) total = total + current.price;
     });
 
+    // @ts-ignore
```
Тут также при типизации контекста уберется и эта ошибка

besovadevka commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -1,81 +1,135 @@
+                    } else {
+                        // @ts-ignore
+                        const isLocked =  (isTop || isBottom) && (current.type === "bun");
+                        //console.log (current);
```
Закомментированный код нужно удалить, так очень важно держать код как можно более чистым

besovadevka commented:
Работа очень хорошая!

Ипользуются контексты, создание которых вынесесно в отдельный файл и папку, все работает корректно) Супер, что используется TS

На что можно стоит внимание:
- вынести папку `services` в папку `src`
- при создании контекстов типизировать их `React.createContext<{ items: ..., setItems: ... }>();` - это поможет убрать большую часть ошибок с TS
- удалить закомментированный код, так как это может в дальнейшем только запутать


Так держать! :)

yuriydyuster committed:
Исправлено:
- очищен код от закомменченных блоков
- типизированны контексты
- скорректированна струткра папок
- константа и сервисные функции вынесены в отдельные модули
- уменьшена область видимости контекста до минимально необходимой

besovadevka commented:
Супер, проделана большая работа!
Все комменты пофикшены)

yuriydyuster commented:
@besovadevka спасибо!))
